packages/aggregator/src/analytics/aggregateAnalytics/adjustOptionsToAggregationList.js

```diff
@@ -0,0 +1,134 @@
+};
+
+const getAdjustedDateRange = (dateRange, aggregationList) =>
+  aggregationList.reduce((currentRange, aggregation) => {
+    const getDateRange = AGGREGATION_TYPE_TO_DATE_RANGE_GETTER[aggregation.type];
+    return getDateRange ? getDateRange(currentRange, aggregation.config) : currentRange;
+  }, dateRange);
+
+const shouldFetchDataSourceEntities = ({ config: { dataSourceEntityType } = {} } = {}) =>
+  dataSourceEntityType;
+
+const shouldFetchRelations = ({
+  type,
+  config: { aggregationEntityType, dataSourceEntityType } = {},
+} = {}) => aggregationEntityType && dataSourceEntityType && !(type === AGGREGATION_TYPES.RAW);
+
+const getAdjustedOrganisationUnitsAndAggregations = async (
+  context,
+  fetchOptions,
+  aggregationList,
+) => {
+  const { hierarchy, organisationUnitCodes } = fetchOptions;
+  if (!aggregationList.some(shouldFetchDataSourceEntities)) {
+    return [organisationUnitCodes, aggregationList];
+  }
+
+  const { session } = context;
+
+  let adjustedOrganisationUnitCodes;
+  const adjustedAggregations = [];
+  const entityConnection = new EntityConnection(session);
+  await Promise.all(
+    aggregationList.map(async aggregation => {
+      if (!shouldFetchDataSourceEntities(aggregation)) {
+        adjustedAggregations.push(aggregation);
+        return;
+      }
+
+      if (!adjustedOrganisationUnitCodes && !shouldFetchRelations(aggregation)) {
+        const { dataSourceEntityType, dataSourceEntityFilter } = aggregation.config;
+        adjustedOrganisationUnitCodes = await entityConnection.getDataSourceEntities(
+          hierarchy,
+          organisationUnitCodes,
+          dataSourceEntityType,
+          dataSourceEntityFilter,
+        );
+        adjustedAggregations.push(aggregation);
+        return;
+      }
+
+      const {
+        aggregationEntityType,
+        dataSourceEntityType,
+        dataSourceEntityFilter,
+      } = aggregation.config;
+      const [
+        dataSourceEntities,
+        relations,
+      ] = await entityConnection.getDataSourceEntitiesAndRelations(
+        hierarchy,
+        organisationUnitCodes,
+        aggregationEntityType,
+        dataSourceEntityType,
+        dataSourceEntityFilter,
+      );
+
+      if (!adjustedOrganisationUnitCodes) {
+        adjustedOrganisationUnitCodes = dataSourceEntities;
+      }
+      adjustedAggregations.push({
+        ...aggregation,
+        config: { ...aggregation.config, orgUnitMap: relations },
+      });
+    }),
+  );
+
+  return [adjustedOrganisationUnitCodes, adjustedAggregations];
+};
+
+/**
+ * Each aggregation in the list may require the original fetch option dimensions (dates, org units etc)
+ * to be adjusted, eg when we need to take into account past/future data
+ *
+ * @param {Object[]} aggregationList
+ */
+export const adjustOptionsToAggregationList = async (context, fetchOptions, aggregationOptions) => {
+  const { aggregations: aggregationList = [] } = aggregationOptions;
```

Comment:
At some point, we should refactor `Aggregator` to expect that
```ts
type AggregationOptions = Aggregation[];
```

While currently, it is
```ts
type AggregationOptions = { aggregation: Aggregation[] };
```

probably because in the past `AggregationOptions` could also be a simple string (or something like that).

There are some utils in `indicators` which allow for polymorphic definition of aggregation. The client can then specify aggregation in various formats (even a simple string), which would be "expanded" to their full `AggregationOptions` format in the main entry to `Aggregator`.

Reply:
Yeah I know what you mean... personally I'm a bit torn since I actually quite like our polymorphic format for aggregations (it's nice to be able to pass in either a single or array or either strings or objects). 

I feel maybe Aggregator should be responsible for handling all those different type of input but formatting them into a consistent shape immediately and we use that throughout the rest of Aggregator?

Reply:
I definitely like polymorphic input as an option! I actually spend some time making indicators work with various aggregation formats. As you say, we should "expand" them once in Aggregator, or once in each package (since `indicators` also benefit from a consistent internal format)